### ISSUE TITLE
[CPU] High per-dispatch latency for small transformer inference

### ISSUE THREAD

**Issue by dbrll** [iree-org/iree#23714]
Running a small GPT-2 model token-by-token on `local-task`, each forward pass takes ~50ms on Apple M5. The same model runs at 2.6ms/token with PyTorch + Accelerate (vecLib/cblas). That's a **~20x gap**.

Single-thread benchmark shows 124ms (no thread pool overhead), so the gap is primarily in the generated code, not dispatch scheduling. The forward compiles to 101 dispatches, averaging ~1.2ms each. Each dispatch costs nearly as much as PyTorch's entire forward pass.

With MPS, the same IREE-compiled model matches PyTorch (6.4ms/tok vs 13ms/tok). GPU pipelining masks the per-dispatch cost that dominates on CPU.

### Benchmarks

```
iree-benchmark-module \
  --module=gpt-forward-batch-1.vmfb \
  --function=gpt_forward \
  --device=local-task \
  --task_topology_group_count=1 \
  --input=1x256xf32 --input=65x384xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=1152x384xf32 \
  --input=384x384xf32 --input=384xf32 --input=384xf32 \
  --input=1536x384xf32 --input=384x1536xf32 --input=384xf32 \
  --input=256x384xf32 --input=65x384xf32 \
  --input="1xi8=01" --input="1xi64=0" --input="1xf32=0.0" \
  --input="1xi64=0" --input="1xi64=0" --input="1xi64=0" --input="1xi64=0" \
  --benchmark_repetitions=5 --time_unit=ms
```

```
# Multi-thread (default, ~8 cores):
BM_gpt_forward/process_time/real_time_mean    50.3 ms    295 ms    5    20.28/s
BM_gpt_forward/process_time/real_time_median  47.3 ms    282 ms    5    21.16/s

# Single-thread (--task_topology_group_count=1):
BM_gpt_forward/process_time/real_time_mean     124 ms    124 ms    5     8.03/s
BM_gpt_forward/process_time/real_time_median   124 ms    124 ms    5     8.07/s
```

The forward compiles to 101 dispatches.

Compiled with `--iree-llvmcpu-target-cpu=host --iree-llvmcpu-enable-ukernels=all` (only improves -17% over no flag). Ukernels likely don't trigger for these small matmul sizes.

VMFB: [gpt-forward-batch-1.vmfb.zip](https://github.com/user-attachments/files/25852824/gpt-forward-batch-1.vmfb.zip)

For comparison:

| Backend            | PyTorch           | IREE               | Ratio      |
| ------------------ | ----------------- | ------------------ | ---------- |
| CPU (`local-task`) | 264ms (2.6ms/tok) | 5540ms (50ms/tok)  | ~20x worse |
| Metal GPU          | 1297ms (13ms/tok) | 1080ms (6.4ms/tok) | 20% better |

### Model details

- NanoGPT with 6 transformer layers, dim=384, 6 heads, FFN=1536
- 10.65M parameters, f32
- Forward only
- Compiled from StableHLO via iree-compile to `embedded-elf-arm_64`

### Analysis

The model is compute-light (small matmuls: 256×384, 384×1152, etc.). For these sizes:
- Compute is cheap (~45M FLOPs per matmul, ~270M total per forward)
- 101 dispatches at ~1.2ms each (single-thread). Each dispatch costs nearly as much as PyTorch's entire forward (2.6ms)
- Single-thread is 124ms vs 50ms multi-thread, so threading helps ~2.5x but the base cost per dispatch is the dominant factor
- PyTorch calls Accelerate/vecLib `cblas_sgemm` single optimized BLAS call per matmul, no tiling/dispatch overhead

I am aware of the high input count, however even accounting for argument marshaling, the 1.2ms per-dispatch latency seems disproportionate.

### Environment

- Apple M5, macOS
- IREE at HEAD (commit bef4552d)

**Comment by hanhanW** [comment]
It is better to try with `--iree-opt-data-tiling` flag, see https://iree.dev/guides/deployment-configurations/cpu/#compile-and-run-a-program

**Comment by dbrll** [comment]
Thanks for the suggestion. I tested `--iree-opt-data-tiling` on a benchmark suite covering micro and macro workloads (CPU, single-thread, Apple M5):

| Benchmark | Baseline | Data-tiling | Delta |
|-----------|----------|-------------|-------|
| matmul [512x512] | 2.520ms | 2.440ms | -3% |
| matmul [1024x1024] | 19.15ms | 18.66ms | -3% |
| gelu [100000] | 5.220ms | 5.170ms | -1% |
| softmax [32x1000] | 4.940ms | 4.910ms | -1% |
| softmax [32x50000] | 7.370ms | 7.450ms | +1% |
| layer-norm [4x256x768] | 6.500ms | 6.570ms | +1% |
| cross-entropy [1x256x50257] | 24.81ms | 25.04ms | +1% |
| value-and-grad [4x512]->[512x512] | 7.000ms | 7.290ms | +4% |
| GPT-2 124M forward | 1190ms | 1200ms | +1% |
| GPT-2 124M value-and-grad | 1410ms | 1280ms | -9% |

All changes are within noise except value-and-grad (-9%). The flag has no measurable effect on the GPT-2 forward pass. The dominant cost remains per-dispatch overhead (~1.2ms each × 101 dispatches), not compute.
